sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -758,7 +758,10 @@ final void moveToOutputFiles(
       }
       // During a failure case, files may have been deleted in an earlier step. Thus
       // we ignore missing files here.
+      long startTime = System.nanoTime();
       FileSystems.rename(srcFiles, dstFiles, StandardMoveOptions.IGNORE_MISSING_FILES);
+      long endTime = System.nanoTime();
+      LOG.info("Renamed {} files in {} seconds.", srcFiles.size(), (endTime - startTime) / 1e9);
```

Comment:
Is this needed? I don't see it in Jenkins logs. It seems to not affect performance much but if it's not needed I'd skip this (I don't know the broader context here). 

If it **is** needed, you can make it more concise and skip all the math using the `Stopwatch` from guava, like this: 

```
Stopwatch stopwatch = Stopwatch.createStarted();
FileSystems.rename(srcFiles, dstFiles, StandardMoveOptions.IGNORE_MISSING_FILES);
LOG.info("Renamed {} files in {} seconds.", srcFiles.size(), stopwatch.stop().elapsed(TimeUnit.SECONDS));
```

Reply:
I'm going to change the code that does batch copies and I needed a way to verify that there are no performance regressions. This PR should do that, but if there is a regression this log line could tell me if the batch operation is slower.
This log line does appear in the terminal if I run the test directly using `gradlew`. Is there a way to export this metric to the dashboard?

Reply:
I don't know an easy way yet. After a better look I noticed that we see only Perfkit logs in Jenkins dashboard and logs from Gradle (that is run by perfkit) are omitted. This is a separate issue imo. 

Are you fine for now with having this printed only locally? 

If you're not fine then: 
 1. we should fix logs in ioits first
 2. use Metrics API and publish this metric to a separate BQ table. This would be concise with future plans regarding IO tests - we plan to leverage metrics api for all metrics instead of relying on perfkit timers only. See LoadTest.java for similar approach (time and total bytes collection and then publishing to BQ): https://github.com/apache/beam/blob/bd5bbf9e1eca68f4f743da4060846f2df27d81f5/sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/LoadTest.java#L78

Reply:
I've done some investigation on Friday.
1. A correction: the log doesn't appear in the terminal - you have to go to Stackdriver logs. Having a history of values is important to detect and investigate regressions, and Stackdriver logs are not kept around for long.
Regardkess, I created a bug to log gradle output in perfkit tests: https://issues.apache.org/jira/browse/BEAM-6251
1. It is possible to report metrics from the workers and query them from the code that launches the pipeline, such as TextIOIT.java. These metrics could be logged and then picked up by https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/blob/master/perfkitbenchmarker/linux_benchmarks/beam_integration_benchmark.py using `regex_util.ExtractAllMatches` and added to the benchmark results (using `sample.Sample` https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/blob/2fc792d137181f2b273fdb7842a3cf513b273e4d/perfkitbenchmarker/linux_benchmarks/beam_integration_benchmark.py#L163). Parsing log files with regexes feels hacky though. I prefer your solution of writing metrics to BQ, and I'd rather write all results (incl. `run_time`) to BQ from TextIOIT and ignore Perfkit's reporting. WDYT?

Reply:
Great investigation! Thanks! 

> I'd rather write all results (incl. run_time) to BQ from TextIOIT and ignore Perfkit's reporting

I fully agree with you. As I mentiond using Metrics API in IOITs is the think that I wanted to do either way it will be there even better if you do this in your test. :)  This should be easy to reuse code from test-utlils module that was used in LoadTest.java I mentioned above. More specifically I mean: 
- `MetricsReader`: that will let you query metrics really easily
- `BigQueryResultsPublisher`: that will let you publish them to BQ. 

As for collecting run_time, I think you can simply use [TimeMonitor](https://github.com/apache/beam/blob/90ef4eb68abca0f904c6b3087aec14b11058b1b5/sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/metrics/TimeMonitor.java) class. It's in load-test module now but with your change it perfectly makes sense to move it to test-utils.

Reply:
> incl. run_time

FWIW, the run_time metric you get from the pipeline will be a more precise one because it won't include Dataflow cluster warmup/teardown time but only will only measure how long the actual test pipeline run time took. This is also a big improvement so it's two birds with one stone.

Reply:
I've added another commit. Tomorrow I'll continue with publishing to BQ.